src/hypertable_restrict_info.c

```diff
@@ -293,18 +368,66 @@ hypertable_restrict_info_add_op_expr(HypertableRestrictInfo *hri, PlannerInfo *r
 							   &lefttype,
 							   &righttype);
 
-	return dimension_restrict_info_add(dri, strategy, c);
+	values = const_values(c);
+	return dimension_restrict_info_add(dri, strategy, values, c->consttype, use_or);
+}
+
+static List *
+get_scalar_op_expr_values(Const *c)
+{
+	ArrayIterator iterator = array_create_iterator(DatumGetArrayTypeP(c->constvalue), 0, NULL);
+	Datum		elem = (Datum) NULL;
+	bool		isnull;
+	List	   *values = NIL;
+
+	while (array_iterate(iterator, &elem, &isnull))
+		if (!isnull)
+			values = lappend(values, DatumGetPointer(elem));
+	return values;
+}
+
+static List *
+get_op_expr_values(Const *c)
+{
+	List	   *values = NIL;
+
+	return lappend(values, DatumGetPointer(c->constvalue));
 }
 
 static void
 hypertable_restrict_info_add_restrict_info(HypertableRestrictInfo *hri, PlannerInfo *root, RestrictInfo *ri)
 {
+	bool		added = false;
+
 	Expr	   *e = ri->clause;
 
-	if (!IsA(e, OpExpr))
+	/* Same as constraint_exclusion */
+	if (contain_mutable_functions((Node *) e))
 		return;
 
-	if (hypertable_restrict_info_add_op_expr(hri, root, (OpExpr *) e))
+	switch (nodeTag(e))
+	{
+		case T_OpExpr:
+			{
+				OpExpr	   *op_expr = (OpExpr *) e;
+
+				added = hypertable_restrict_info_add_expr(hri, root, op_expr->args, op_expr->opno, get_op_expr_values, false);
+				break;
+			}
+
+		case T_ScalarArrayOpExpr:
+			{
+				ScalarArrayOpExpr *scalar_expr = (ScalarArrayOpExpr *) e;
+
+				added = hypertable_restrict_info_add_expr(hri, root, scalar_expr->args, scalar_expr->opno, get_scalar_op_expr_values, scalar_expr->useOr);
```

Comment:
another possible way to simplify this code: isn't it true that if scalar_expr->useOr == false, we can just iterate over the list here and call dimension_restrict_info_add on each individual list element. This is because conceptually ` a op 'id1' and a op 'id2'` is the same as `a op ALL('id1' and 'id2')` but right now these take vastly different code paths. Ideally, I think `dimension_restrict_info_add` should only take lists of items that correspond to "OR" lists. For AND lists we can just call dimension_restrict_info_add in a loop with the elements in the list.

Reply:
I am not sure if for certain cases we can directly call `dimension_restrict_info_add ` here b/c there are couple of checks we need to run before ( eg. various checks in `hypertable_restrict_info_add_expr`)
I wouldn't say that `a op 'id1' AND a op 'id2'` and `a op ALL('id1','id2')` take vastly different code paths: the only difference between these two is how you extract values from `Const`.